### PR TITLE
feat: implement not

### LINKED ISSUE #18: feat: implement 0x19 - NOT opcode
---
fork: Frontier
group: Comparison & Bitwise Logic Operations
---

*Index 1 is top of the stack. See [PUSH](https://www.evm.codes/#60).*

## Stack input

0. `a`: binary value.

## Stack output

0. `~a`: the bitwise NOT result.

## Example

| * | Input | Output |
|--:|------:|-------:|
| `1` | `0` | `0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF` |

[Reproduce in playground](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='PUSH1%200%5CnNOT'_).

## Error cases

The state changes done by the current context are [reverted](https://www.evm.codes/?fork=merge#FD) in those cases:
- Not enough values on the stack.

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

resolves #18 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/instructions/comparison_operations.cairo [hunk 1]
```diff
@@ -53,7 +55,19 @@ impl ComparisonAndBitwiseOperations of ComparisonAndBitwiseOperationsTrait {
     /// 0x19 - NOT
     /// Bitwise NOT operation
     /// # Specification: https://www.evm.codes/#19?fork=shanghai
-    fn exec_not(ref self: ExecutionContext) {}
+    fn exec_not(ref self: ExecutionContext) {
+        let maybe_a = self.stack.pop();
+        let mut result: u256 = 0;
+        match maybe_a {
+            Option::Some(a) => {
+                result = ~a;
```
Comment (ftupas) [inline src/instructions/comparison_operations.cairo:63]: TIL this is an operator 🤯

File: src/tests/test_instructions/test_comparison_operations.cairo [hunk 2]
```diff
@@ -52,3 +53,57 @@ fn test_xor_half_same_pair() {
...
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(
+        ctx.stack.peek().unwrap() == BoundedInt::<u256>::max(), 'stack top should be 0xFFF..FFFF'
+    );
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_max_uint() {
```
Comment (ftupas) [inline src/tests/test_instructions/test_comparison_operations.cairo:77]: nit: we can add `not` so it's `test_not_max_uint`

**Review by ftupas — APPROVED** [review]
Looks good! Just nit on test name

File: src/instructions/comparison_operations.cairo [hunk 3]
```diff
@@ -53,7 +55,19 @@ impl ComparisonAndBitwiseOperations of ComparisonAndBitwiseOperationsTrait {
...
+            },
+            Option::None => {
+                panic_with_felt252(STACK_UNDERFLOW);
```
Comment (enitrat) [inline src/instructions/comparison_operations.cairo:66]: > The state changes done by the current context are reverted in those cases:

> Not enough gas.
> Not enough values on the stack.

This is not just about your implementation but about the global implementation of other opcodes / pop_n method: only the current context is supposed to revert if there are not enough values on the stack; not the whole tx

File: src/tests/test_instructions/test_comparison_operations.cairo (hunk 2 above)
Comment (enitrat) [reply src/tests/test_instructions/test_comparison_operations.cairo:77]:
```suggestion
fn test_not_max_uint() {
```

File: src/instructions/comparison_operations.cairo (hunk 3 above)
Comment (Eikix) [reply src/instructions/comparison_operations.cairo:66]: Can you create a feature issue / could be considered a bugfix, but I think we need to think on the architecture side how to handle this!

**Review by enitrat — APPROVED** [review]

**Review by enitrat — APPROVED** [review]

**Pull request merged**